File: Source/Utility/Hooks/use-once.hook.ts
import { Janitor as Maid } from "@rbxts/janitor";
import { useEffect } from "@rbxts/roact-hooked";

export const useOnce = (Callback: () => void | (() => void)) => useEffect(() => { return Callback(); }, []);

export function useJanitor<Output, Cached extends void | object = void>(Callback: (Garbage: Maid<Cached>) => Output) {
	const Janitor = new Maid<Cached>();

    useOnce(() => { Callback(Janitor); return () => Janitor.Destroy(); });
};

File: Output/Utility/Hooks/use-once.hook.lua
-- Compiled with roblox-ts v2.2.0
local TS = _G[script]
local Maid = TS.import(script, TS.getModule(script, "@rbxts", "janitor").src).Janitor
local useEffect = TS.import(script, TS.getModule(script, "@rbxts", "roact-hooked").src).useEffect
local useOnce = function(Callback)
	return useEffect(function()
		return Callback()
	end, {})
end
local function useJanitor(Callback)
	local Janitor = Maid.new()
	useOnce(function()
		Callback(Janitor)
		return function()
			return Janitor:Destroy()
		end
	end)
end
local default = useOnce
return {
	useJanitor = useJanitor,
	useOnce = useOnce,
	default = default,
}
